draft-ietf-tls-esni.md

```diff
@@ -309,33 +303,100 @@ This protocol also defines the "esni_required" alert, which is sent by the
 implementations to track, alongside each PSK established by a previous
 connection, whether the connection negotiated this extension with the
 "esni_accept" response type. If so, this is referred to as an "ESNI PSK".
 Otherwise, it is a "non-ESNI PSK". This may be implemented by adding a new field
 to client and server session states.
 
-## Client Behavior {#client-behavior}
+## Incorporating Outer Extensions {#outer-extensions}
+
+Some TLS 1.3 extensions can be quite large
+and having them both in the inner and outer ClientHello wil lead to
+a very large overall size. One particularly pathological example
+is "key_share" with post-quantum algorithms. In order to reduce
+the impact of duplicated extensions, the client may use the
+"outer_extension" extension.
 
-### Sending an encrypted SNI {#send-esni}
+~~~
+   enum {
+       esni_extension(TBD), (65535)
+   } ExtensionType;
 
-In order to send an encrypted SNI, the client MUST first select one of
-the server ESNIKeyShareEntry values and generate an (EC)DHE share in the
-matching group. This share will then be sent to the server in the
-"encrypted_server_name" extension and used to derive the SNI encryption key. It does not affect the
-(EC)DHE shared secret used in the TLS key schedule. The client MUST also select
-an appropriate cipher suite from the list of suites offered by the
-server. If the client is unable to select an appropriate group or suite it
-SHOULD ignore that ESNIConfig value and MAY attempt to use another value provided
-by the server. The client MUST NOT send encrypted SNI using groups or cipher suites
-not advertised by the server.
+   struct {
+       ExtensionType extension;
+       uint8 hash<32..255>;
+   } OuterExtension;
+~~~~
+
+This extension MUST only be used in ClientHelloInner and contains
+a digest of the corresponding extension in ClientHelloOuter.
+When sending ClientHello, the client first computes ClientHelloInner,
+including the PSK binders, and then MAY substitute any extensions
+which it knows will be duplicated in ClientHelloOuter with
+the corresponding "outer_extension". The hash value is computed
+over the entire extension, including the type and length field
+and uses the same hash as for the KDF used to encrypt ClienHelloInner.
+This process is reversed by client-facing server upon receipt.
+
+Clients SHOULD only use this mechanism for extensions which are
+large. All other extensions SHOULD appear in both ClientHelloInner
+and ClientHelloOuter even if they have identical values.
+
+Multiple "outer_extension" extensions MAY appear in a ClientHelloInner
+(this is a violation of normal TLS rules, but the resulting ClientHelloInner
+is never processed directly). However, there MUST NOT be
+multiple "outer_extension" extensions with the same extension code point.
+
+
+# Client Behavior {#client-behavior}
+
+## Sending an encrypted ClientHello {#send-esni}
+
+In order to send an encrypted SNI, the client MUST first generate its
```

Comment:
I've noticed that this draft uses "encrypted SNI" and "ESNI" interchangeably, but unpredictably. Would recommend taking a pass over all of these acronyms once the naming is settled ("ECHO" etc...)